Add sequencing API and use it for PVW & PGM source and Transition Style

arikorn opened the pull request:
Add sequencing API and use it for PVW & PGM source and Transition Style
- user specifies the set of sources to cycle through
- choose either **sequential** order, **'random sets'** (no repeat until all sources have been selected), **'fully random'** (may repeat any time, aka selection with replacement).
- Transition Style is modified as an example showing a second way to use this API for actions with only 3-4 choices (no random sequences, just add a choice to the dropdown menu).

This addresses issues #21 and #126 (#126 in specific will be addressed in a subsequent PR)

## Some details:
Buttons defined with these actions can use the corresponding variables as their label text, so the button shows the current value.

The "sequencing" API is contained in util.ts.

It consists of a dropdown menu (`sequenceOrderDropdown`) and a function to be embedded in the action's callback (`nextInSequence`)

`nextInSequence` is designed with two modes in mind:

1. Long Sequences with options for random selection: --  Create a new action with a multidropdown contain all of the options. `default:` should be set to all of the options. -- Add `sequenceOrderDropdown`as the next element in the `options:` array -- Define the action callback something like:
~~~
  callback: async (action) => {
        const srcSequence = action.options.Sources as number[]
        const newSource = nextInSequence(srcSequence, state.PgmSrc, action)

        await sendCommand(ActionId.PvwIndex, ReqType.Set, [newSource])
    }
~~~
The first arg is the multidropdown selections, the second arg is the current state of the item.

2. Short (3-4 item) sequences, sequential option only -- Add something like `{ id: -1, label: 'Toggle' }` to the dropdown choices. -- optionally add a multidropdown with all available choices, as above. -- add code similar to:
~~~
  callback: async (action) => {
    	let choice = getOptNumber(action, 'TransitionStyle')
    	if (choice === -1) {
    		// Toggle: cycle through all available choices sequentially:
    		const sizes = action.options.TransitionStyleSequence as number[]
    		const curStyle = state.selectTransitionStyle.style
    		choice = nextInSequence(sizes, curStyle) // default order is sequential.
    	}
    	await sendCommand(ActionId.TransitionIndex, ReqType.Set, [choice])
    },
~~~

# visual examples:
# **_Type 1: (PVW Source)_**
![image](https://github.com/user-attachments/assets/725ecbc8-78ee-48c0-91fd-bda19e547239)


# **_Type 2: (Transition Type_**)
 note that the second pulldown is hidden if toggle isn't selected)
![image](https://github.com/user-attachments/assets/35d77f92-0a75-4f76-934b-4685aae91b38)

fa1k3n commented:
Two things:

* I don't think its needed for two separate actions for PGM and PVW, just make the bus a dropdown selection 

* To make this truly competitive with other systems I think it needs a timer function to auto switches between the sources, at least that is what I meant with issue #21. It is also missing some of the behaviors from that issue. So this PR will not solve #21 completely . (This bullet can be addressed now or later. I'm intending to add a timer server where you can add callbacks to be called on a regular interval. So when this is done then this issue can be updated to use that and do auto switch)

arikorn commented:
@fa1k3n 
> Two things:
> 
> * I don't think its needed for two separate actions for PGM and PVW, just make the bus a dropdown selection

Two action parallels the existing (separate) actions for PGM and PVW. I think it would be confusing for the users for these actions to be organized differently than those actions.

> * To make this truly competitive with other systems I think it needs a timer function to auto switches between the sources, at least that is what I meant with issue [Add input sequencing support #21](https://github.com/bitfocus/companion-module-osee-gostream/issues/21). It is also missing some of the behaviors from that issue. So this PR will not solve [Add input sequencing support #21](https://github.com/bitfocus/companion-module-osee-gostream/issues/21) completely .

I agree, it doesn't fully solve either of the two Issues I referenced, nor did I say it does! However, I believe the PR does stand on its own merit, and establishes a solid foundation for addressing both of those issues.

  As for your specific desire to implement a timer, I've already addressed issues and solutions to that problem in my comments in issue #21. Briefly: (1) you can't make a module action that takes more than 5 seconds and (2) the proper way to do automated sequences in Companion is through triggers. Please see my comments there for further details. Regardless, I don't think implementing an automated timer should hold up this PR, and my reading of your comment is that you feel the same way.

fa1k3n commented:
Ok for the first comment , I understand your point and perhaps also the original set pgm and pvw actions should be consolidated into one too . But that's for another PR 

For the second point I think you should add the timing from the start . There is no problem having timers longer than 5 seconds as long as they don't keep the action open , which they will not do in this case . As for the timer server I don't know why you reference that as I have edited that out long before you read the issue , I did that because I changed my mind and I don't think this is needed anymore

arikorn commented:
@fa1k3n this PR is about creating a "compact console" such as shown below and in PR #147 (merged) and mentioned indirectly in issue #136, both PR and issue are closed now, so I didn't reference them. My apologies.

Timers are not required nor even desired for the "compact console." (TBH, random is not needed either for this PR, I just added it as a "bonus" to make this PR more "enticing" or at least provide a foundation for your issue #21 while solving a difficult issue requested there, since it is related to sequencing. As an "unintended benefit", users can easily create automated sequences with triggers using this PR -- I even made a video about a simpler case in the GoStream FB group a while back. -- However, again, automated/timer actions is not the intention of this PR) As best as I understand WRT to coding and other hurdles, timers are a completely separate functionality which should be its own PR and be debated elsewhere, if necessary. Please let's make this PR about supporting the  "compact console," which I have found to be extremely handy in my work!!!

Thank you.

![image](https://github.com/user-attachments/assets/6f16f5fd-039f-4f87-b4e8-71f35b71038b)
